project-api/ballerina-projects/src/main/java/io/ballerina/projects/writers/BaloWriter.java

```diff
@@ -29,18 +59,347 @@
+            for (File moduleDir : directoryListing) {
+                if (moduleDir.isDirectory()) {
+                    // add module
+                    Path moduleDirInBalo = packageInBalo.resolve(defaultPackageName + "." + moduleDir.getName());
+                    Files.createDirectory(moduleDirInBalo);
+
+                    // copy resources directory
+                    copyResourcesDir(moduleDir.toPath(), moduleDirInBalo);
+                    // only add .bal files in the module root
+                    copyBallerinaSrcFiles(moduleDir.toPath(), moduleDirInBalo);
+                }
+            }
+        }
+    }
+
+    private static void copyResourcesDir(Path sourceDir, Path sourceDirInBalo) throws IOException {
+        Path resourcesDir = sourceDirInBalo.resolve("resources");
+        Path resourcesSrcDir = sourceDir.resolve("resources");
+
+        // if resources not exists ignore copying
+        if (resourcesSrcDir.toFile().exists()) {
+            File[] resourceFiles = new File(String.valueOf(resourcesSrcDir)).listFiles();
+            if (resourceFiles != null && resourceFiles.length > 0) {
+                Files.walkFileTree(resourcesSrcDir, new CopyResources(resourcesSrcDir, resourcesDir));
+            }
+        }
+    }
+
+    private static void copyBallerinaSrcFiles(Path sourceDirPath, Path targetPath) throws IOException {
+        File[] defaultModuleFiles = new File(String.valueOf(sourceDirPath)).listFiles();
+        if (defaultModuleFiles != null) {
+            for (File file : defaultModuleFiles) {
+                if (file.isFile() && file.getName().endsWith(".bal")) {
+                    Files.copy(file.toPath(), targetPath.resolve(file.getName()), StandardCopyOption.REPLACE_EXISTING);
+                }
+            }
+        }
+    }
+
+    private static void addPlatformLibs(Path root, Path projectDirectory, BallerinaToml ballerinaToml)
+            throws IOException {
+        //If platform libs are defined add them to balo
+        List<Library> platformLibs = ballerinaToml.getPlatform().libraries;
+        if (platformLibs == null) {
+            return;
+        }
+        Path platformLibsDir = root.resolve("lib");
+        Files.createDirectory(platformLibsDir);
+
+        for (Library lib : platformLibs) {
+            if (lib.getModules() == null && lib.getScope() == null) {
+                Path libPath = Paths.get(lib.getPath());
+                Path nativeFile = projectDirectory.resolve(libPath);
+                Path libFileName = libPath.getFileName();
+                if (libFileName == null) {
+                    continue;
+                }
+                Path targetPath = platformLibsDir.resolve(libFileName.toString());
+                try {
+                    Files.copy(nativeFile, targetPath, StandardCopyOption.REPLACE_EXISTING);
+                } catch (IOException e) {
+                    throw new BLangCompilerException("Dependency jar not found : " + lib.toString());
```

Comment:
Files.copy can throw an I/O exception for reasons other than for a file not found. Therefore, it would be better if we make it more generic and append the exception `e` to the message.
```suggestion
                    throw new BLangCompilerException("Error while trying to add platform library to the BALO: " + lib.toString(), e);
```

Reply:
Added